File: LeetCode/42. Trapping Rain Water.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0

        res = 0
        H = len(height)
        # Arrays to store the maximum height to the left and right of each element
        leftMax = [0] * H
        rightMax = [0] * H

        leftMax[0] = height[0]
        for i in range(1, H):
            leftMax[i] = max(leftMax[i - 1], height[i])

        rightMax[H - 1] = height[H - 1]
        for i in range(H - 2, -1, -1):
            rightMax[i] = max(rightMax[i + 1], height[i])

        for i in range(H):
            res += min(leftMax[i], rightMax[i]) - height[i]

        return res


s = Solution()
print(s.trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]))

File: 2. LeetCode/42. Trapping Rain Water.cpp
class Solution {
public:
    int trap(const std::vector<int> &height) {
        if (height.empty()) {
            return 0;
        }

        int H = height.size();
        int res = 0;

        std::vector<int> leftMax(H);
        std::vector<int> rightMax(H);

        // 왼쪽에서부터 최대 높이 저장
        leftMax[0] = height[0];
        for (int i = 1; i < H; ++i) {
            leftMax[i] = std::max(leftMax[i - 1], height[i]);
        }

        // 오른쪽에서부터 최대 높이 저장
        rightMax[H - 1] = height[H - 1];
        for (int i = H - 2; i >= 0; --i) {
            rightMax[i] = std::max(rightMax[i + 1], height[i]);
        }

        // 물을 잡을 수 있는 양 계산
        for (int i = 0; i < H; ++i) {
            res += std::min(leftMax[i], rightMax[i]) - height[i];
        }

        return res;
    }
};